### DISCUSSION TITLE
Access JMX metrics from outside K8s

### DISCUSSION THREAD

**Discussion by alonpr** [strimzi/strimzi-kafka-operator#7469] [Q&A]
Hi,

When setting `jmxOptions: {}` it exposes JMX metrics (on port 9999) which are accessible only from other pods but not from outside k8s (due to ephemeral rmi port...) - Is that correct?

**Comment by scholzj** [comment]
I don't know JMX much. If you would want to expose it to outside your cluster, you will probably need to create some load balancer, node port or similar services for it to expose it. Strimzi does not expose it to the outside on its own and IMHO the JMX security is relatively weak so I would not do it.

**Reply by alonpr** [reply]
To follow up on this, does `jmxOptions: {}` set both JMX_PORT and KAFKA_JMX_OPTS ?

**Reply by scholzj** [reply]
I don't know. You would need to check the code or try it out. It does what is needed to enable the JMX.

**Reply by alonpr** [reply]
It's set in `set_kafka_jmx_options.sh` ->
`KAFKA_JMX_OPTS="-Dcom.sun.management.jmxremote.port=9999 -Dcom.sun.management.jmxremote.rmi.port=9999 -Dcom.sun.management.jmxremote=true -Djava.rmi.server.hostname=$(hostname -i) -Djava.net.preferIPv4Stack=true"`

Is it possible to override `KAFKA_JMX_OPTS` somehow that instead of this value it'd be something that I provide?

**Reply by scholzj** [reply]
I don't know - I don' think so.

**Comment by omersiar** [comment]
This would also help troubleshoot JVM further deeply with JMX RMI tools.

Currently default behaviour for strimzi to set java.rmi.server.hostname JMX configuration to Pod's IP address, which makes impossible to connect JMX RMI interface from outside world (behind NAT or Port-Forwarded proxy like kubectl)

If this is intended (to be only routable within the kubernetes network) then jmxOptions should expose a configuration for those who want change this behaviour enabling us to start a troubleshooting session.
https://strimzi.io/docs/operators/latest/configuring.html#type-KafkaJmxOptions

If it does not advertise Pod IP or advertise 127.0.0.1 or localhost then this would not be a problem.  -javaagent would not be affected since it is also another local process.

**Comment by omersiar** [comment]
Putting it behind a Loadbalancer would not work because once initial connection is made, RMI advertise the IP address of the Pod, unless modified by the loadbalancer (I could not able to find a way to do it other than https://github.com/marcin-wolak/SimpleRmiDiscoverer?tab=readme-ov-file#simplermidiscoverer) client would try to connect that IP.

**Comment by omersiar** [comment]
> Hi,
> 
> When setting jmxOptions: {} it exposes JMX metrics (on port 9999) which are accessible only from other pods but not from outside k8s (due to ephemeral rmi port...) - Is that correct?

Correct, but not because it is ephemeral, Strimzi explicitly set RMI port to as same as JMX > 9999, it is not ephemeral.

**Comment by dberardo-com** [comment]
any news here ? documentation on this is so scarce ...

the load balancer techniques would not work then ? i feel not even the kubbectl port forward works ... this is so bad ...

**Reply by omersiar** [reply]
Yep,  
We could have also point it to a name and then client would be redirected to routable IP which can be localhost (kubectl port-forward works this way)
https://github.com/strimzi/strimzi-kafka-operator/pull/9939

With current setup, I could not find any JMX client (other than the pentest tools) not honouring the advertised IP address.

**Reply by katheris** [reply]
@mimaison are you able to provide any help here?

**Reply by mimaison** [reply]
Not entirely sure how it works with Strimzi. In plain Kafka, you can set the `JMX_PORT` or `KAFKA_JMX_OPTS` environment variable to configure the JMX port and set JMX options. Then you need to expose that port with whatever Kubernetes Service mechanism (NodePort, LoadBalancer) you use in your environment.

As Jakub said above, JMX has pretty weak security so it's typically not recommended, and hence not documented.
